### PR TITLE
test: moving waitForCounter calls off of using sleep

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: n/a (test only)
Testing: many tsan runs
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by alyssawilk** [comment]
@jmarantz feel free to ignore this until next week given you're already looking at the accessor PR - you just seem like the right reviewer for this one.
If we like it, I can move gauge lookups over at some point

File: test/integration/server.h [hunk 1]
```diff
@@ -299,10 +354,19 @@ class IntegrationTestServer : public Logger::Loggable<Logger::Id::testing>,
              Server::FieldValidationConfig validation_config, uint32_t concurrency);
 
   void waitForCounterEq(const std::string& name, uint64_t value) override {
-    TestUtility::waitForCounterEq(statStore(), name, value, time_system_);
+    while (statsAllocator().getCounter(name) == nullptr ||
+           statsAllocator().getCounter(name)->value() != value) {
+      absl::MutexLock l(&statsAllocator().mutex());
+      statsAllocator().condvar().Wait(&statsAllocator().mutex());
```
Comment (jmarantz) [inline test/integration/server.h:360]: quick drive-by; can you just make this waitFor functionality part of the Allocator interface? Also is there a real need to put the counter/gauge impls in the header? I kind of like having them in the CC file. It wasn't obvious to me why they had to be in the header.

Comment (jmarantz) [reply test/integration/server.h:360]: oh I see...you subclassed it. I actually think you might not need to and can just use absl::Mutex::Await in the impl without adding any new real functionality to the prod path.

Comment (jmarantz) [reply test/integration/server.h:360]: thinking further I realize this comment is wrong. we don't have a mutex guarding counter values so you need a counter subclass to get a notifier on incrementing those.

i still wish we did not have to put so much in the header, but maybe we can figure out something.

also note you don't need condvars at all; you can just use mutexes and Await.

Comment (alyssawilk) [reply test/integration/server.h:360]: Yeah, so I started switching over to await, but I think it's a bit harder to read.  WDYT?  

   while (statsAllocator().getCounter(name) == nullptr ||
           statsAllocator().getCounter(name)->value() != value) {
      AwaitData data = std::make_tuple(&statsAllocator(), &name, value);
      statsAllocator().mutex().Await(absl::Condition(+[](AwaitData* data) {
        Stats::NotifyingAllocatorImpl* allocator =  std::get<0>(*data);
        const std::string& name =  *std::get<1>(*data);
        uint64_t value = std::get<2>(*data);
        return allocator->getCounter(name) != nullptr && allocator->getCounter(name)->value() == value;
      }, &data));
    }
  }

File: test/integration/server.h [hunk 2]
```diff
@@ -298,11 +352,31 @@ class IntegrationTestServer : public Logger::Loggable<Logger::Id::testing>,
              bool defer_listener_finalization, ProcessObjectOptRef process_object,
              Server::FieldValidationConfig validation_config, uint32_t concurrency);
 
+  using AwaitData = std::tuple<Stats::NotifyingAllocatorImpl*, const std::string*, uint64_t>;
   void waitForCounterEq(const std::string& name, uint64_t value) override {
-    TestUtility::waitForCounterEq(statStore(), name, value, time_system_);
+    while (statsAllocator().getCounter(name) == nullptr ||
+           statsAllocator().getCounter(name)->value() != value) {
+      AwaitData data = std::make_tuple(&statsAllocator(), &name, value);
+      statsAllocator().mutex().Await(absl::Condition(+[](AwaitData* data) {
```
Comment (jmarantz) [inline test/integration/server.h:360]: oh yeah I have an outstanding PR which provides a modern wrapper for Await to hide this messiness.

File: test/integration/server.h [hunk 3]
```diff
@@ -151,6 +152,59 @@ class TestScopeWrapper : public Scope {
...
+    {
+      absl::MutexLock l(&mutex_);
+      counters_.emplace(counter->name(), counter);
+    }
+    return counter;
+  }
+
+  // Allow getting the counter directly from the allocator, since it's harder to
+  // signal when the counter has been added to a given stats store.
+  virtual Stats::CounterImpl* getCounter(std::string name) {
```
Comment (jmarantz) [inline test/integration/server.h:192]: You could also make this call block until the counter is created via an await. There's a bunch of similar code in TestStore in test/common/stats/stat_test_utility.h -- it'd be nice to re-use that somehow.

Also I'm wondering how you'd feel about wrapping and delegating rather than changing the underlying allocator...it's a bit more work because you'd have to making forwarding methods but also less intrusive to prod code.

From a practical perspective I have a preference for not putting significant prod impls in headers as it makes debugging a little messier as the code gets replicated in multiple object files, but it's not a deal breaker. There are other ways around that though.

At a minimum call this getCounterFromString per convention in that file.

File: test/integration/server.h [hunk 4]
```diff
@@ -298,11 +352,31 @@ class IntegrationTestServer : public Logger::Loggable<Logger::Id::testing>,
              bool defer_listener_finalization, ProcessObjectOptRef process_object,
              Server::FieldValidationConfig validation_config, uint32_t concurrency);
 
+  using AwaitData = std::tuple<Stats::NotifyingAllocatorImpl*, const std::string*, uint64_t>;
   void waitForCounterEq(const std::string& name, uint64_t value) override {
-    TestUtility::waitForCounterEq(statStore(), name, value, time_system_);
+    while (statsAllocator().getCounter(name) == nullptr ||
```
Comment (jmarantz) [inline test/integration/server.h:357]: WDYT of blocking first on the counter existing, and then blocking on the value?

File: test/integration/server.h [hunk 5]
```diff
@@ -298,11 +352,31 @@ class IntegrationTestServer : public Logger::Loggable<Logger::Id::testing>,
...
+        const std::string& name =  *std::get<1>(*data);
+        uint64_t value = std::get<2>(*data);
+        return allocator->getCounter(name) != nullptr && allocator->getCounter(name)->value() == value;
+      }, &data));
+    }
   }
 
   void waitForCounterGe(const std::string& name, uint64_t value) override {
+    while (statsAllocator().getCounter(name) == nullptr ||
+           statsAllocator().getCounter(name)->value() < value) {
```
Comment (jmarantz) [inline test/integration/server.h:371]: You could mostly factor this out with the above based on a lambda predicate.

File: test/integration/server.h (hunk 3 above)
Comment (jmarantz) [reply test/integration/server.h:192]: I realize I've hinted at "other ways around this" and should just say what I mean:

allocator_impl.h is somewhat broadly included, transitively. I think we can have allocator_impl_private.h which has these inline template methods in them, and include those only in allocator_impl.cc in prod, plus in a new cc file implementing your notiying allocator.

But LMK what you know about delegating in TestStore (test/common/stats/stat_test_utility.h), rather than subclassing the allocator for this. Maybe that won't be the right integration point for integration tests?

Comment (alyssawilk) [reply test/integration/server.h:192]: I don't think that would work.
At the end of the day we need to know when the counter has been incremented.  AFIK that can only be done by either refactoring the actual production code to increment through the store (rather than getting the counter and incrementing directly) or subclassing the counter.

Comment (alyssawilk) [reply test/integration/server.h:192]: Will look into wrapping and keeping the base classes in the .cc file.  If I do that work would you be OK with condvar for now, and you could clean it up once your mutex await fix lands?

File: test/integration/server.h (hunk 2 above)
Comment (alyssawilk) [reply test/integration/server.h:360]: Ok if I revetr this while that lands?  it was a WIP and again I think it decreases readability.

**Comment by jmarantz** [comment]
going back to condvar for now sgtm

**Comment by alyssawilk** [comment]
Ok, had to do the dreaded force push to unwind the half-done counter PR, but honestly given the stats move back to the cc file may be easier to look at all diffs rather than the delta (which is from your first round of review)

**Comment by alyssawilk** [comment]
heh. the lack of sleeps appears to be causing some flakes where there were unnoticed races.  Will upload when I have those sorted out.

**Comment by jmarantz** [comment]
It's awesome how going into the deterministic world shines a light on the messy dark corners :)

Looking forward to the update and thanks for doing this!

**Comment by alyssawilk** [comment]
Should be deflaked - PTAL when you get a chance

**Comment by jmarantz** [comment]
I'll take a look but something in this PR has significantly increased memory usage for stats, which is why 'released' failed in CI:
```
   Which is: 52702
  44491
```

File: test/integration/server.h [hunk 6]
```diff
@@ -152,6 +153,80 @@ class TestScopeWrapper : public Scope {
...
+  const SymbolTable& constSymbolTable() const override { return counter_->constSymbolTable(); }
+
+private:
+  std::unique_ptr<Stats::Counter> counter_;
+  absl::Mutex& mutex_;
+  absl::CondVar& condvar_;
+};
+
+// A stats allocator which creates NotifyingCounters rather than regular CounterImpls.
+class NotifyingAllocatorImpl : public Stats::AllocatorImpl {
```
Comment (jmarantz) [inline test/integration/server.h:193]: oh I wonder if this notifier is being used in the StatsIntegrationTest now.

File: test/integration/stats_integration_test.cc [hunk 7]
```diff
@@ -337,6 +338,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithRealSymbolTable) {
   // 2020/05/05  10908    36345       36800   router: add InternalRedirectPolicy and predicate
   // 2020/05/13  10531    36537       36800   Refactor resource manager
   // 2020/05/20  11223    36603       36800   Add primary clusters tracking to cluster manager.
+  // 2020/05/27  11289    36603       44846   Test-only change wrapping counters in test counters
```
Comment (jmarantz) [inline test/integration/stats_integration_test.cc:341]: hmmm...maybe we can make these memory tests not use the notifying allocator. I'd like to keep these numbers prod accurate if possible.

File: test/integration/server.h [hunk 8]
```diff
@@ -433,6 +521,8 @@ class IntegrationTestServerImpl : public IntegrationTestServer {
   Stats::Store* stat_store_{};
   Network::Address::InstanceConstSharedPtr admin_address_;
   absl::Notification server_gone_;
+  Stats::SymbolTablePtr symbol_table_;
+  Stats::NotifyingAllocatorImpl stats_allocator_;
```
Comment (jmarantz) [inline test/integration/server.h:525]: How about plumbing a ctor param to allow integration tests to choose between production stats and suppporting waitaForCounter*?

File: test/integration/stats_integration_test.cc (hunk 7 above)
Comment (alyssawilk) [reply test/integration/stats_integration_test.cc:341]: We can, but we'd have to plumb the boolean through several layers of code special case every waitFor function plus do a bunch of dynamic casts in the server to switch.  Is it that much worse to have a known fixed increase?

**Comment by jmarantz** [comment]
IMO yes: the purpose of this test is to measure memory overhead and I think we should try to do it faithfully. Is it a huge amount of work to plumb some bools through the constructors?

File: test/integration/stats_integration_test.cc (hunk 7 above)
Comment (jmarantz) [reply test/integration/stats_integration_test.cc:341]: IMO yes: the purpose of this test is to measure memory overhead and I think we should try to do it faithfully. Is it a huge amount of work to plumb some bools through the constructors?

Comment (jmarantz) [reply test/integration/stats_integration_test.cc:341]: another alternative is to pull the stats memory tests out of the integration-test framework. I'm not sure how difficult that would be.

Comment (alyssawilk) [reply test/integration/stats_integration_test.cc]: Pulling it out would be more work.
I'm just going to claim if you need real stats, the counter (and eventually gague) functions don't work.  If someone needs to check them they can implement the ugly workarounds.

Comment (jmarantz) [reply test/integration/stats_integration_test.cc]: Yeah I looked at that and I agree that it's not worth it to divest it from the BaseIntegrationTest.

I'm not following you about the "if you need real stats" comment; the only function of this test is to track and understand per-cluster and per-host memory overhead of stats. 

Another option which might make it easier is to have an optional virtual method in BaseIntegrationTest which you could override to ask for real stats. That would be less toil than plumbing in a bool to ctors. WDYT?

Comment (alyssawilk) [reply test/integration/stats_integration_test.cc]: I'm not convinced it's less toil, and given I already did the bool-toil I'd prefer to keep this as-is?

Comment (jmarantz) [reply test/integration/stats_integration_test.cc]: oh sorry didn't realize you already did this with bools; I'll take a quick look. thanks.

File: test/integration/server.h [hunk 9]
```diff
@@ -152,6 +153,80 @@ class TestScopeWrapper : public Scope {
...
+      if (counter->name() == "cluster_manager.cluster_removed") {
+      }
+      condvar_.Signal();
+    }
+    return counter;
+  }
+
+  // Allow getting the counter directly from the allocator, since it's harder to
+  // signal when the counter has been added to a given stats store.
+  virtual Stats::Counter* getCounterLockHeld(std::string name) EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/server.h]: did 3 of the 4 - I'd prefer the wait to be in one place rather than (while (waitForCounter()->value !=) { wait }

**Comment by jmarantz** [comment]
needs format fix.

**Comment by jmarantz** [comment]
Thinking about this more, there is no reason to wait for the counter to
exist. You can just create if needed, and return the notifying version.

The advantage is you would avoid exposing the class mutex and condvar.
Wdyt?

**Comment by alyssawilk** [comment]
creating the stat correctly requires the  tag_extracted_name and stat_name_tags which aren't available there.  I'd prefer to not shave more yaks here to avoid a second wait.

**Comment by mattklein123** [comment]
Merging master should fix CI.

/wait
